### PR TITLE
Eliminate nest.flatten() call in _filtered_call

### PR DESCRIPTION
**Pull request by jonathanchu33**
Redundant, following nest.flatten() in `_convert_numpy_outputs`.
Note: PR not ready for merge yet, submitted for @sun51 and @kkimdev to check progress
Edit: In review

### CODE REVIEW

File: tensorflow/python/eager/function.py [hunk 1]
```diff
@@ -3191,10 +3199,16 @@ def _create_graph_function(self, args, kwargs, override_flat_arg_shapes=None):
...
+  def _define_function_with_shape_relaxation(self,
+                                             args,
+                                             kwargs,
+                                             flat_args,
+                                             flat_kwargs):
     """Define a function, relaxing arg shapes to avoid unnecessary retracing."""
+    flat_args_all = nest.flatten((args, kwargs), expand_composites=False)
+
     any_composite_args = any(isinstance(x, composite_tensor.CompositeTensor)
-                             for x in nest.flatten((args, kwargs)))
```
[the remark this answers is not in the document]
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py:3197]: Yes, this is true - `flat_args_all` is defined with `expand_composites` False, see line 3208

File: tensorflow/python/eager/function.py [hunk 2]
```diff
@@ -2796,7 +2803,7 @@ def format_error_message(inputs, input_signature):
         flat_sequence=flatten_inputs,
         expand_composites=True)
 
-  return inputs
+  return inputs, flatten_inputs
```
[the remark this answers is not in the document]
Comment (jonathanchu33) [reply tensorflow/python/eager/function.py]: Done

**Comment by jonathanchu33** [comment]
Benchmarks - mean & [median of means] of 10 trials with 30000 iters each:
defun_matmul: 185/184 us -> 171/167
defun_matmul_async: 93/92 us -> 89/88 us
defun_matmul_with_signature: 198/193 us -> 189/184 us

File: tensorflow/python/eager/def_function.py [hunk 3]
```diff
@@ -910,14 +910,16 @@ def fn_with_cond(*inner_args, **inner_kwds):
...
+                            inner_flat_args, inner_flat_kwds))
 
     # We've created variables and are unable to lift the initialization graphs,
     # so we fall back to initializing with conds while running the function.
-    canon_args, canon_kwds = \
+    canon_args, canon_kwds, flat_args, flat_kwds = \
         self._stateful_fn._function_spec.canonicalize_function_inputs(  # pylint: disable=protected-access
             *args, **kwds)
-    return function_lib.defun(fn_with_cond)(*canon_args, **canon_kwds)
+    # TODO(jlchu): verify that modification to fn_with_cond works
```
Comment (sun51) [inline tensorflow/python/eager/def_function.py:920]: please verify before submission, once verified please remove this comment

File: tensorflow/python/eager/function.py [hunk 4]
```diff
@@ -3191,10 +3199,16 @@ def _create_graph_function(self, args, kwargs, override_flat_arg_shapes=None):
         shared_func_graph=False)
     return graph_function
 
-  def _define_function_with_shape_relaxation(self, args, kwargs):
+  def _define_function_with_shape_relaxation(self,
+                                             args,
+                                             kwargs,
+                                             flat_args,
+                                             flat_kwargs):
```
Comment (sun51) [inline tensorflow/python/eager/function.py:3206]: is this still needed?

Comment (jonathanchu33) [reply tensorflow/python/eager/function.py:3203]: It's not absolutely required, but there is still a call to `nest.flatten` with expanded composites in L3247/3261 that it replaces

File: tensorflow/python/eager/def_function.py (hunk 3 above)
Comment (jonathanchu33) [reply tensorflow/python/eager/def_function.py]: Removed, per discussion that we would verify with GitHub checks passing

**Comment by jonathanchu33** [comment]
Pending tests passing, the last step would be to fix relevant aspects of `tensorflow/python/autograph/g3doc/reference/error_handling.md`

File: tensorflow/python/eager/function.py [hunk 5]
```diff
@@ -3241,14 +3255,16 @@ def _define_function_with_shape_relaxation(self, args, kwargs):
...
           expand_composites=True)
 
     graph_function = self._create_graph_function(
         args, kwargs, override_flat_arg_shapes=relaxed_arg_shapes)
     self._function_cache.arg_relaxed[rank_only_cache_key] = graph_function
 
-    return graph_function, args, kwargs
+    return (graph_function,
+            nest.flatten(args, expand_composites=True),
+            nest.flatten(kwargs, expand_composites=True))
```
Comment (sun51) [inline tensorflow/python/eager/function.py:3267]: this is same as flat_args, right?

Comment (sun51) [inline tensorflow/python/eager/function.py:3267]: same as flat_kwargs, right?

Comment (jonathanchu33) [reply tensorflow/python/eager/function.py:3267]: It seems that args and kwargs can be modified/repacked in the `if any_composite_args` case (L3242/3256). Will check if they end up flattening to the same thing though

**Review by sun51 — APPROVED** [review]

**Comment by jonathanchu33** [comment]
Last step is to fix relevant parts of `tensorflow/python/autograph/g3doc/reference/error_handling.md`

**Pull request merged**